File: solutions/3903-smallest-stable-index-i/prefix-sum.cpp
// time: O(n)
// space: O(n)

#include <algorithm>
#include <vector>

using namespace std;

class Solution {
  public:
    int firstStableIndex(vector<int> &nums, int k) {
        const int n = nums.size();
        int suffix[n];
        int mx = 0;

        suffix[n - 1] = nums[n - 1];
        for (int i = n - 2; i >= 0; i--) suffix[i] = min(suffix[i + 1], nums[i]);

        for (int i = 0; i < n; i++) {
            mx = max(mx, nums[i]);
            if (mx - suffix[i] <= k) return i;
        }

        return -1;
    }
};

File: solutions/3903-smallest-stable-index-i/prefix-sum.js
// time: O(n)
// space: O(n)

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const firstStableIndex = (nums, k) => {
    const n = nums.length;
    const suffix = new Array(n).fill(Infinity);
    let max = 0;

    suffix[n - 1] = nums[n - 1];

    for (let i = n - 2; i >= 0; i--)
        suffix[i] = Math.min(suffix[i + 1], nums[i]);

    for (let i = 0; i < n; i++) {
        max = Math.max(max, nums[i]);
        if (max - suffix[i] <= k) return i;
    }

    return -1;
};
